flaml/automl/model.py

```diff
@@ -386,6 +418,342 @@ def config2params(self, config: dict) -> dict:
+        self.model_classes_ = None
+        self.model_n_classes_ = None
+
+    def fit(self, X_train, y_train=None, budget=None, free_mem_ratio=0, **kwargs):
+        start_time = time.time()
+        if self.model_n_classes_ is None and self._task not in ["regression", "rank"]:
+            self.model_n_classes_, self.model_classes_ = len_labels(
+                y_train, return_labels=True
+            )
+        df_train = self._preprocess(X_train, y_train)
+        n_iter = self.params.get(self.ITER_HP, self.DEFAULT_ITER)
+        trained = False
+        mem0 = psutil.virtual_memory().available if psutil is not None else 1
+        _kwargs = kwargs.copy()
+        # TODO: update regression model and rank model, update ParamList_LightGBM_
+        if self._task not in ["regression", "rank"]:
+            if "objective" not in _kwargs:
+                _kwargs["objective"] = (
+                    "binary" if self.model_n_classes_ == 2 else "multiclass"
+                )
+        for k in list(_kwargs.keys()):
+            if k not in self.estimator_params:
+                _kwargs.pop(k)
+        if (
+            (not self._time_per_iter or abs(self._train_size - df_train.count()) > 4)
+            and budget is not None
+            or self._mem_per_iter < 0
+            and psutil is not None
+        ) and n_iter > 1:
+            self.params[self.ITER_HP] = 1
+            self._t1 = self._fit(df_train, **_kwargs)
+            if budget is not None and self._t1 >= budget or n_iter == 1:
+                return self._t1
+            mem1 = psutil.virtual_memory().available if psutil is not None else 1
+            self._mem1 = mem0 - mem1
+            self.params[self.ITER_HP] = min(n_iter, 4)
+            self._t2 = self._fit(df_train, **_kwargs)
+            mem2 = psutil.virtual_memory().available if psutil is not None else 1
+            self._mem2 = max(mem0 - mem2, self._mem1)
+            self._mem_per_iter = min(self._mem1, self._mem2 / self.params[self.ITER_HP])
+            self._time_per_iter = (
+                (self._t2 - self._t1) / (self.params[self.ITER_HP] - 1)
+                if self._t2 > self._t1
+                else self._t1
+                if self._t1
+                else 0.001
+            )
+            self._train_size = df_train.count()
+            if (
+                budget is not None
+                and self._t1 + self._t2 >= budget
+                or n_iter == self.params[self.ITER_HP]
+            ):
+                # self.params[self.ITER_HP] = n_iter
+                return time.time() - start_time
+            trained = True
+        if n_iter > 1:
+            max_iter = min(
+                n_iter,
+                int(
+                    (budget - time.time() + start_time - self._t1) / self._time_per_iter
+                    + 1
+                )
+                if budget is not None
+                else n_iter,
+                int((1 - free_mem_ratio) * mem0 / self._mem_per_iter)

```

Comment:
This uses memory consumption information to decide the maximum iterations. For the spark training, the memory consumption in the head node doesn't indicate the consumption in the worker nodes. So this part can be removed.

Reply:
Removed.